common/quotas/global/algorithm/requestweighted.go

```diff
@@ -0,0 +1,351 @@
+		// account for missed updates
+		reduce, gc := missedUpdateScalar(snap.now.Sub(history.lastUpdate), snap)
+		if gc {
+			// old, clean up
+			delete(ih, id)
+			continue
+		}
+
+		actual := HostWeight((history.accepted + history.rejected) * reduce)
+
+		weights[id] = actual // populate with the reduced values so it doesn't have to be calculated again
+		total += actual      // keep a running total to scale all values when done
+		usedRPS += history.accepted * reduce
+	}
+
+	if len(ih) == 0 {
+		// completely empty Limit, gc it as well
+		delete(a.usage, key)
+		return nil, 0
+	}
+
+	for id := range ih {
+		// scale by the total.
+		// this also ensures all values are between 0 and 1 (inclusive)
+		weights[id] = weights[id] / total
+	}
+	return weights, usedRPS
+}
+
+func (a *impl) HostWeights(host Identity, limits []Limit) (weights map[Limit]HostWeight, usedRPS map[Limit]PerSecond) {
+	a.mut.Lock()
+	defer a.mut.Unlock()
+
+	weights = make(map[Limit]HostWeight, len(limits))
+	usedRPS = make(map[Limit]PerSecond, len(limits))
+	snap := a.snapshot()
+	for _, lim := range limits {
+		hosts, used := a.getWeightsLocked(lim, snap)
+		if len(hosts) > 0 {
+			usedRPS[lim] = used // limit is known, has some usage
+			if weight, ok := hosts[host]; ok {
+				weights[lim] = weight // host has a known weight
+			}
+		}
+	}
+	return weights, usedRPS
+}
+
+func (a *impl) GC() Metrics {
+	a.mut.Lock()
+	defer a.mut.Unlock()
+
+	m := Metrics{}
+	snap := a.snapshot()
+	// TODO: too costly? can check the first-N% each time and it'll eventually visit all keys, demonstrated in tests.
+	for lim, dat := range a.usage {
+		for host, hist := range dat {
+			if _, gc := missedUpdateScalar(snap.now.Sub(hist.lastUpdate), snap); gc {
+				// clean up stale host data within limits
+				delete(dat, host)
+				m.RemovedHostLimits++
+			} else {
+				m.HostLimits++
+			}
+		}
+
+		// clean up stale limits
+		if len(dat) == 0 {
+			delete(a.usage, lim)
+			m.RemovedLimits++
+		} else {
+			m.Limits++
+		}
+	}
+
+	return m
+}
+
+// returns a snapshot of config and "now" for easier chaining through calls, and reducing calls to
+// non-trivial field types like dynamic config.
+func (a *impl) snapshot() configSnapshot {
+	return configSnapshot{
+		now:       a.clock.Now(),
+		weight:    a.cfg.NewDataWeight(),
+		rate:      a.cfg.UpdateRate(),
+		maxMissed: a.cfg.MaxMissedUpdates(),
+		gcAfter:   a.cfg.GcAfter(),
+	}
+}
+
+// missedUpdateScalar returns an amount to multiply old RPS data by, to account for missed updates outside SLA.
+func missedUpdateScalar(elapsed time.Duration, cfg configSnapshot) (scalar PerSecond, gc bool) {
+	reduce := PerSecond(1.0)
+
+	// fast path: check the bounds for "new enough to ignore"
+	if elapsed <= time.Duration(float64(cfg.rate)*cfg.maxMissed) {
+		return reduce, false // within SLA, no changes
+	}
+	// fast path: check the bounds for "old enough to prune"
+	if elapsed >= cfg.rate*time.Duration(cfg.gcAfter) {
+		return reduce, true
```

Comment:
why return 1.0 in this case? shouldn't it be 0?

Reply:
0 would be safer, yeah, I'll do that.

I think I just didn't think about it because all gc cases are/should be ignoring the value entirely... but 0 makes that correct either way 👍